Adding VPA modifications tests for the issue #93 in autoscaler

kallurbsk opened the pull request:
Tests:
TestTotalSamplesCountMemorySamplesOnly
    - Issue with covering TotalSamplesCount adding only CPU samples and with memory samples not being counted.
TestRecordOOMSampleAndCountTotalSamples
    - Issue with recording OOM Samples but still TotalSamplesCount not being updated in aggregate container state.
TestRecommendationForZeroTotalSamplesCount
    - Issue with not getting any recommendation if the CPU samples are not added and there by TotalSamplesCount is 0.
TestAddMemorySurgeWithOOMBumpUpRatio
    - Hardcoded OOMBumpUpRatio (=1.2) and hence takes more time to recover with more number of restarts. Can be made user defined.
TestAddMemorySurgeWithOOMMinBump
    - Hardcoded OOMMinBump (=100MB) and hence takes more time to recover with more number of restarts. Can be made user defined.
TestRecordCPUThrottlingInformation
    - Showcases the need of containerState.RecordOOM like function for recording CPU throttle information which should be added to CPU histogram if the CPU usage is constantly above current pod limits set
TestRecommenderScaleUp
    - Normal scale up behavior of the existing VPA recommender
TestRecommenderScaleDown
    - Normal scale down behavior of the existing VPA recommender

The above test cases currently fail indicating the problem
with to be fixed as mentioned in #93. They validate the
change required in the behavior for VPA recommendation.

**What this PR does / why we need it**:
The above set of UT validates the issues in #93 
We need to fix the existing VPA recommender code to make
sure we have proper behavior which is blocking us at specific places.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       improvement|noteworthy|action
- target_group:   user|operator|developer
-->
```improvement operator

```

kallurbsk committed:
Adding VPA modifications tests for the issue #93 in autoscaler

Tests:
TestTotalSamplesCountMemorySamplesOnly
TestRecordOOMSampleAndCountTotalSamples
TestRecommendationForZeroTotalSamplesCount
TestAddMemorySurgeWithOOMBumpUpRatio
TestAddMemorySurgeWithOOMMinBump
TestRecordCPUThrottlingInformation

The above test cases currently fail indicating the problem
with to be fixed as mentioned in #93. They validate the
change required in the behavior for VPA recommendation.

amshuman-kr commented on vertical-pod-autoscaler/pkg/recommender/routines/recommender_routines_test.go:
```diff
@@ -0,0 +1,267 @@
+	return cluster.AddSample(&sample)
+}
+
+func addTestCPUHistogram(samples []float64, weight float64, testTimeStamp time.Time, options util.HistogramOptions) util.Histogram {
+	cpuHistogram := util.NewHistogram(options)
+	for _, sample := range samples {
+		cpuHistogram.AddSample(sample, weight, testTimestamp)
+	}
+
+	return cpuHistogram
+}
```
There are a couple of issues with this function.

1. It couples construction of the histogram object with the adding of samples. There might be cases where you want to add samples to an existing histogram object. Please take the histogram as an additional parameter (the caller can construct it if needed).
1. The function takes multiple sample values but single weight and timestamp. There will certainly be cases where the samples must be of different weights and timestamps. Either make struct (say `resourceSample` which have three fields like `value`, `weight` and `timestamp`) and take a slice of that struct as a parameter in this function.

kallurbsk replied:
ACK. 
1. Decoupled and created a separate functions for same. 
2. Created a separate struct and filling it separately.

amshuman-kr commented on vertical-pod-autoscaler/pkg/recommender/routines/recommender_routines_test.go:
```diff
@@ -0,0 +1,267 @@
+	return cpuHistogram
+}
+
+func addTestMemoryHistogram(samples []float64, weight float64, timeStamp time.Time, options util.HistogramOptions) util.Histogram {
+	memoryHistogram := util.NewHistogram(options)
+	for _, sample := range samples {
+		memoryHistogram.AddSample(sample, weight, testTimestamp)
+	}
+
+	return memoryHistogram
+}
```
Same comments as for `addTestCPUHistogram`.

kallurbsk replied:
ACK.

amshuman-kr commented on vertical-pod-autoscaler/pkg/recommender/routines/recommender_routines_test.go:
```diff
@@ -0,0 +1,267 @@
+	return resourceEstimation, surgeSamplesCount
+}
+
+func GetPercentileResourceEstimation(cpuPercentile float64, memPercentile float64,
+	cpuHistogram util.Histogram, memoryHistogram util.Histogram) model.Resources {
+
+	estimator := logic.NewPercentileEstimator(cpuPercentile, memPercentile)
+
+	return estimator.GetResourceEstimation(
+		&model.AggregateContainerState{
+			AggregateCPUUsage:    cpuHistogram,
+			AggregateMemoryPeaks: memoryHistogram,
+		})
+}
```
Do we need this function? Can't we call [podResourceRecommender.GetRecommendedPodResources](https://github.com/gardener/autoscaler/blob/a842944a648b7e3d87b2915f4c5fea077db9969b/vertical-pod-autoscaler/pkg/recommender/logic/recommender.go#L56:34) instead?

kallurbsk replied:
ACK. Removed as not using estimator but using GetRecommendedPodResources now.

amshuman-kr commented on vertical-pod-autoscaler/pkg/recommender/routines/recommender_routines_test.go:
```diff
@@ -0,0 +1,267 @@
+			memorySample = fn(memorySample)
+		}
+		memoryPeaksHistogram.AddSample(memorySample, 1.0, testTimestamp)
+		resourceEstimation = GetPercentileResourceEstimation(cpuPercentile, memPercentile,
+			cpuHistogram, memoryPeaksHistogram)
```
Can't we call [podResourceRecommender.GetRecommendedPodResources](https://github.com/gardener/autoscaler/blob/a842944a648b7e3d87b2915f4c5fea077db9969b/vertical-pod-autoscaler/pkg/recommender/logic/recommender.go#L56:34) instead?

kallurbsk replied:
ACK. Removed as not using estimator but using GetRecommendedPodResources now.

amshuman-kr commented on vertical-pod-autoscaler/pkg/recommender/routines/recommender_routines_test.go:
```diff
@@ -0,0 +1,267 @@
+	memoryPeaksHistogram := addTestMemoryHistogram(memSamples, 1.0, testTimestamp, config.MemoryHistogramOptions)
+
+	ResourcePercentile := 0.9 // Same as in podResourceRecommender for target estimator
+	resourceEstimation := GetPercentileResourceEstimation(ResourcePercentile, ResourcePercentile,
+		cpuHistogram, memoryPeaksHistogram)
```
Can't we call [podResourceRecommender.GetRecommendedPodResources](https://github.com/gardener/autoscaler/blob/a842944a648b7e3d87b2915f4c5fea077db9969b/vertical-pod-autoscaler/pkg/recommender/logic/recommender.go#L56:34) instead?

kallurbsk replied:
ACK. Removed as not using estimator but using GetRecommendedPodResources now.

amshuman-kr commented on vertical-pod-autoscaler/pkg/recommender/routines/recommender_routines_test.go:
```diff
@@ -0,0 +1,267 @@
+
+	// Adding sudden surge sample to memory histogram
+	memoryPeaksHistogram.AddSample(4e10, 1.0, testTimestamp)
+	resourceEstimation = GetPercentileResourceEstimation(ResourcePercentile, ResourcePercentile,
+		cpuHistogram, memoryPeaksHistogram)
```
Can't we call [podResourceRecommender.GetRecommendedPodResources](https://github.com/gardener/autoscaler/blob/a842944a648b7e3d87b2915f4c5fea077db9969b/vertical-pod-autoscaler/pkg/recommender/logic/recommender.go#L56:34) instead?

kallurbsk replied:
ACK. Removed as not using estimator but using GetRecommendedPodResources now.

amshuman-kr commented on vertical-pod-autoscaler/pkg/recommender/routines/recommender_routines_test.go:
```diff
@@ -0,0 +1,267 @@
+	memoryPeaksHistogram := addTestMemoryHistogram(memSamples, 1.0, testTimestamp, config.MemoryHistogramOptions)
+
+	ResourcePercentile := 0.9 // Same as in podResourceRecommender for target estimator
+	resourceEstimation := GetPercentileResourceEstimation(ResourcePercentile, ResourcePercentile,
+		cpuHistogram, memoryPeaksHistogram)
```
Can't we call [podResourceRecommender.GetRecommendedPodResources](https://github.com/gardener/autoscaler/blob/a842944a648b7e3d87b2915f4c5fea077db9969b/vertical-pod-autoscaler/pkg/recommender/logic/recommender.go#L56:34) instead?

kallurbsk replied:
ACK. Removed as not using estimator but using GetRecommendedPodResources now.

amshuman-kr commented on vertical-pod-autoscaler/pkg/recommender/routines/recommender_routines_test.go:
```diff
@@ -0,0 +1,267 @@
+
+	// Adding sudden surge sample to memory histogram
+	memoryPeaksHistogram.AddSample(4e10, 1.0, testTimestamp)
+	resourceEstimation = GetPercentileResourceEstimation(ResourcePercentile, ResourcePercentile,
+		cpuHistogram, memoryPeaksHistogram)
```
Can't we call [podResourceRecommender.GetRecommendedPodResources](https://github.com/gardener/autoscaler/blob/a842944a648b7e3d87b2915f4c5fea077db9969b/vertical-pod-autoscaler/pkg/recommender/logic/recommender.go#L56:34) instead?

kallurbsk replied:
ACK. Removed as not using estimator but using GetRecommendedPodResources now.

amshuman-kr commented on vertical-pod-autoscaler/pkg/recommender/routines/recommender_routines_test.go:
```diff
@@ -0,0 +1,267 @@
+	memoryPeaksHistogram := addTestMemoryHistogram(memSamples, 1.0, testTimestamp, config.MemoryHistogramOptions)
+
+	ResourcePercentile := 0.9 // Same as in podResourceRecommender for target estimator
+	resourceEstimation := GetPercentileResourceEstimation(ResourcePercentile, ResourcePercentile,
```
Can't we call [podResourceRecommender.GetRecommendedPodResources](https://github.com/gardener/autoscaler/blob/a842944a648b7e3d87b2915f4c5fea077db9969b/vertical-pod-autoscaler/pkg/recommender/logic/recommender.go#L56:34) instead?

kallurbsk replied:
ACK. Removed as not using estimator but using GetRecommendedPodResources now.

amshuman-kr commented on vertical-pod-autoscaler/pkg/recommender/routines/recommender_routines_test.go:
```diff
@@ -0,0 +1,267 @@
+// VPA recommender estimation to beat the current surge in usage.
+// The surge is determined and samples w.r.t that is added using
+// model.OOMBumpUpRatio
+func TestAddMemorySurgeWithOOMBumpUpRatio(t *testing.T) {
```
These test case functions follow a pattern. Can you have a helper function to capture and reuse that pattern (along the lines of [template method pattern](https://www.geeksforgeeks.org/template-method-design-pattern/))?

For example,
```go
func doTestRecommender(
        initialCPUSamples, initialMemorySamples resourceSamples,
        assertInitialRecommendations func(RecommendedPodResources),
        doTestAction func(),
        assertPostTestActionRecommendations func(RecommendedPodResources),
) {
       ...
       addTestCPUHistogram(cpuHistogram, initialCPUSamples)
       addTestMemoryHistogram(memoryPeakHistogram, initialCPUSamples)
       ...
       assertInitialRecommendations(recommender.GetRecommendedPodResources(GetContainerNameToAggregateStateMap(vpaObj)))

       doTestActionFn()
       
       assertPostTestActionRecommendations(recommender.GetRecommendedPodResources(GetContainerNameToAggregateStateMap(vpaObj)))
}
```

It can then be called like below.
```go
func TestAddMemorySurgeWithOOMBumpUpRatio(t *testing.T) {
        doTestRecommender(
                []resourceSamples{
                        {1.0, 1.0, timestamp1},
                        {0.9, 1.0, timestamp2},
                },
                []resourceSamples{
                        {1.0, 1.0, timestamp1},
                        {0.9, 1.0, timestamp2},
                },
                func(res RecommendedPodResources) { /* do asserts on initial recommendations */ },
                func() {
                        //This could be direct manipulation of the histograms of indirect manipulation by calling other functions that manipulate the histogram like RecordOOM
                        addTestMemoryHistogram(memoryPeakHistogram, []resourceSamples{
                                 { 1.0, 1.0, timestamp3 },
                        })
                },
                func(res RecommendedPodResources) { /* do asserts on post test action recommendations */ },
        )
}
```

kallurbsk replied:
ACK.
Converted the tests to use Template method pattern as suggested.
Thanks for the reference example!

amshuman-kr reviewed (changes requested):
Thanks for the PR @kallurbsk. I have given my comments below.

kallurbsk committed:
Addressing review comments, and added Scale tests

1. Added Normal Scale up and Scale down tests to the list
2. Addressed Amshuman's comments to move the tests to Template Method
   pattern
3. Modified the tests validated based on GetRecommendedPodResources

kallurbsk committed:
Addressed Amshuman's comments

1. Modified CPU Throttling to get better samples and 3 way comparison
2. Added common resource recommender calculation in doTestRecommender
3. Other minor comments on names and comments

ashwani2k commented:
As discussed with @voelzmo we don't wish to use these on the fork, but rather contribute directly to the upstream. So closing this in favor of efforts taken centrally to address the issues.

/close
